src/transformers/NestedLoopTransformation.cc

```diff
@@ -0,0 +1,214 @@
+/*
+ * Copyright (c) 2024, Konstantin Britikov <britikovki@gmail.com>
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#include "NestedLoopTransformation.h"
+
+#include "QuantifierElimination.h"
+#include "TransformationUtils.h"
+#include "graph/ChcGraph.h"
+#include "utils/SmtSolver.h"
+
+std::unique_ptr<NestedLoopTransformation::WitnessBackTranslator>
+NestedLoopTransformation::transform(ChcDirectedGraph const & graph) {
+    auto vertices = graph.getVertices();
+    auto copyGraph = graph;
+    std::vector<EId> loop = detectLoop(copyGraph);
+    std::vector<WitnessInfo> loops;
+    while (loop.size() > 1) {
+        auto wtnsInfo = copyGraph.contractConnectedVertices(loop);
+        loops.push_back(wtnsInfo);
+        loop = detectLoop(copyGraph);
+    }
+    return std::make_unique<NestedLoopTransformation::WitnessBackTranslator>(graph, std::move(copyGraph), loops);
```

Comment:
That's not what I meant.
The backtranslator does not need to know about the new graph, no?
If you look how other transformations work, they return a pair of unique pointers. One to the transformed graph, one to the backtranslator.
We should use the same design here.

Reply:
Backtranslator needs to know about the new graph though.

It is being reused to produce model from the invalidity Witness, to navigate the new graph and based on it to transform. witness into the new representation

Or do you mean we should pass newly produced graph to the witness translation as one of the arguments, but not as a class member?

Reply:
OK, I see.
Then I suggest that the Backtranslator has a shared pointer to the transformed graph and you also return a shared pointer to the same graph from `translate`.

Reply:
Fixed